### DISCUSSION TITLE
Variable Level Weekly Decomposition

### DISCUSSION THREAD

**Discussion by njprice17** [google/meridian#807] [Q&A]
I'm trying to recover a variable level weekly decomposition so not just a breakdown by media channel and baseline but by each individual variable, would it be possible to share how to do this? 

Thanks!

**Comment by Jon77Ruler** [comment]
To extract weekly baseline estimates, use this:

`mmm_data_analysis = analyzer.Analyzer(mmm)`

```
decomp = mmm_data_analysis.expected_vs_actual_data(confidence_level = 0.9).to_dataframe()
decomp.to_csv(f"basedecomp.csv")
```
You can add np.mean (like below) if you want to just take the mean value.


Then to extract the individual variables, something like this.  Note that you'll need to align your variable names for the headers; the order being media, non-media, control if I recall correctly. Sorry - I've never bothered trying to join headers back in Python but if you work out how just throw it in below?

```
all_weeks_list = data.time.values.tolist()

all_weeks_decomp = mmm_data_analysis.incremental_outcome(
    selected_times = all_weeks_list,
    aggregate_times = False
)

all_weeks_mean = np.mean(all_weeks_decomp, axis=(0,1))
df_allweeks_mean = pd.DataFrame(all_weeks_mean)
df_allweeks_mean.to_csv(f"var_decomposition.csv")
```

**Reply by njprice17** [reply]
Thanks for the response! 

Is it also possible to get out the baseline factors by variable as well?

**Reply by Jon77Ruler** [reply]
I'm unsure what you're asking for, I'm sorry?  The baseline for a media variable with zero spend is zero.  It is only by accruing spend that value is assigned via adstock and saturation curves?

**Comment by cpulavarthi** [comment]
Hello @njprice17,

Thank you for contacting us!

If you are looking the breakdown of the baseline into the impact from various control variables, I would advise against that as controls don’t have causal impact and breaking the baseline down to compute those causal attributions may not make sense. Check our documentation on [reasons controls don’t have causal inference or baseline breakdown](https://developers.google.com/meridian/docs/advanced-modeling/control-variables#rsns) for more details. You may check [Discussion #249](https://github.com/google/meridian/discussions/249) for an approximate workaround if you need it for your analysis and for exploratory purposes.

If you are looking for simple weekly or geo-level decompositions of metrics in the `Analyzer` module, you may get them by setting the arguments `aggregate_times` and `aggregate_geos` to `False` for most methods in the [`Analyzer` module](https://developers.google.com/meridian/reference/api/meridian/analysis/analyzer/Analyzer). For example, the [`incremental_outcome`](https://developers.google.com/meridian/reference/api/meridian/analysis/analyzer/Analyzer#incremental_outcome) method allows you to get the incremental outcome over the baseline for each channel (with or without organic channels depending on whether `include_non_paid_channels=True`). A sample code for the same is as follows:

```
model_analysis = analyzer.Analyzer(meridian)
incremental_tf = model_analysis.incremental_outcome(use_posterior=True, aggregate_times=False, aggregate_geos=False)
```

The `incremental_tf` in the above code will be a tensor of dimensions `(n_chains, n_draws, n_geos, n_times, n_channels)` and you would need to aggregate over the `n_chains` and `n_draws` dimensions using mean to get the final incremental outcome values for each time period, geo and channel. A similar method can be followed with other methods in this module. Please note that the sum of the incremental outcomes from each channel and the baseline attribution will give you the total expected outcome (i.e., the estimated value of the target KPI).

To help me provide the most tailored code, could you please clarify what specific individual variables (beyond media channels) you are looking to decompose? This will allow me to point you to the most appropriate method in Meridian.

Feel free to reach out for any further questions or suggestions regarding Meridian.

Google Meridian Support Team

**Reply by nd-mmm** [reply]
Hey guys - I have used the analyzer.Analyzer(meridian) module to output weekly decomp for media and non-media channels i.e 

`incremental_outcomes = analyzer.compute_incremental_outcome_aggregate(
    use_posterior=True,
    include_non_paid_channels=True,
    aggregate_geos=True,
    aggregate_times=False,      # keep weekly granularity
    inverse_transform_outcome=True
)`

I then aggregate over the number of draws and chains: `mean_incremental_outcome = tf.reduce_mean(incremental_outcomes, axis=[0, 1])`

as I want to produce a % weekly decomp. The issue is the data in the output file is not aligned to the html output from meridian. for example the meridian html suggests 73% incrementality while the output suggests closer to 50% is there a reason these would not be aligned?

**Reply by cpulavarthi** [reply]
Hello @nd-mmm,

Thank you for contacting us!

The Summary report doesn't have any visualization displaying weekly decomposition of percentage contributions of each channel. The channel-wise percentage contribution is displayed in the waterfall chart (3rd visualization) but it is aggregated over the entire time period and doesn't have weekly granularity. Contribution over time is displayed in the area plot (6th visualization) but it isn't percentage contribution though it would contain the same time granularity as your dataset. The units of this contribution chart are in terms of the target KPI (Revenue, sales, ROI, conversions, etc).

The comparable quantities are the contribution over time area plot (6th visualization) and the output from the `compute_incremental_outcome_aggregate()` method as both of them would output channel-wise contributions at weekly granularity in the units of the target KPI. But be sure to notice that the contribution over time plot actually is a stacked area plot, with the blue baseline at the bottom and the difference between the first trendline and the blue area of the baseline would be the actual value of the incremental outcome contribution of the channel which is correspondingly named in the legend, and so on.

I am unable to reproduce the discrepancies you mentioned in my execution using the demo notebook. Please share additional information regarding: 
- Which visualization of the summary output you are comparing with the output of the `compute_incremental_outcome_aggregate()` method's output. 
- It’s not clear what comparison you are trying to make as the code snippet provided is an aggregation of the incremental outcome and is itself not directly comparable to a percentage contribution. What exactly are 73% and 50%?

Please reach out with additional information requested so that I may better understand and assist you with your issues.

Google Meridian Support Team
